cpp/src/dictionary/search.cu

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/dictionary/search.hpp>
+#include <cudf/dictionary/detail/search.hpp>
+
+#include <thrust/copy.h>
+
+namespace cudf
+{
+namespace dictionary
+{
+namespace detail
+{
+
+template<typename Element>
+struct copy_if_fn
+{
+    column_device_view const d_keys;
+    Element const d_key;
+
+    __device__ bool operator()(size_type idx) const
+    {
+        if( d_keys.is_null(idx) )
+            return false;
+        return d_key == d_keys.element<Element>(idx);
+    }
+};
+
+/**
+ * @brief Find index of a given key within a dictionary's keys column.
+ * 
+ * The index is the position within the keys column where the given key (scalar) is found.
+ * The keys column is sorted and unique so only one value is expected.
+ * The result is an integer scalar identifing the index value.
+ * If the key is not found, the resulting scalar has `is_valid()=false`.
+ */
+struct find_index_fn
+{
+    template<typename Element, std::enable_if_t<not std::is_same<Element, dictionary32>::value>* = nullptr>
+    std::unique_ptr<numeric_scalar<int32_t>> operator()( dictionary_column_view const& input, scalar const& key,
+                                                         rmm::mr::device_memory_resource* mr, cudaStream_t stream ) const
+    {
+        auto result = std::make_unique<numeric_scalar<int32_t>>( 0, false, stream, mr );
+        if( input.size()==0 )
+            return result;
+        CUDF_EXPECTS( input.keys().type()==key.type(), "search key type must match dictionary keys type" );
+        auto keys_view = column_device_view::create(input.keys(),stream);
+        auto find_key = static_cast<experimental::scalar_type_t<Element> const&>(key);
+        auto iter = thrust::copy_if( rmm::exec_policy(stream)->on(stream), 
```

Comment:
Why not a thrust::find? I imagine it'd be faster than a copy_if.

Reply:
So `thrust::find()` returns an iterator which then I can diff with to get the index but then I need to copy the integer into device memory via numeric_scalar. The `thrust::copy_if()` does both in one kernel -- find and copy.

Reply:
I expect to find the key most of the time.